// @dynatrace/openkit-js/dist/node/core/beacon/strategies/BeaconCache.js
"use strict";
/*
 * Copyright 2019 Dynatrace LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.BeaconCacheImpl = void 0;
const Utils_1 = require("../../utils/Utils");
class BeaconCacheImpl {
    constructor() {
        this.entries = [];
    }
    register(session, prefix, payloadBuilder, state) {
        const entry = {
            session,
            initialized: false,
            prefix,
            communicationState: state,
            builder: payloadBuilder,
        };
        this.entries.push(entry);
        return entry;
    }
    getAllUninitializedSessions() {
        return this.entries.filter((entry) => !entry.initialized);
    }
    getAllClosedSessions() {
        return this.entries.filter((entry) => entry.initialized && entry.session.isShutdown());
    }
    getAllInitializedSessions() {
        return this.entries.filter((entry) => entry.initialized);
    }
    unregister(entry) {
        Utils_1.removeElement(this.entries, entry);
    }
    getEntriesCopy() {
        return this.entries.slice(0);
    }
}
exports.BeaconCacheImpl = BeaconCacheImpl;
