// @dynatrace/openkit-js/dist/node/core/beacon/strategies/SendingStrategy.js
"use strict";
/*
 * Copyright 2019 Dynatrace LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
class AbstractSendingStrategy {
    init(sender, cache) {
        this.sender = sender;
        this.cache = cache;
        this.afterInit();
    }
    shutdown() {
        return Promise.resolve();
    }
    entryAdded(entry) {
        // stub
    }
    afterInit() {
        // stub
    }
    flush() {
        if (!this.sender) {
            return Promise.resolve();
        }
        return this.sender.flush();
    }
}
exports.AbstractSendingStrategy = AbstractSendingStrategy;
