// @dynatrace/openkit-js/dist/node/core/PlatformConstants.js
"use strict";
/*
 * Copyright 2019 Dynatrace LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.errorTechnologyType = exports.defaultServerId = exports.agentTechnologyType = exports.platformTypeOpenKit = exports.protocolVersion = exports.openKitVersion = void 0;
/**
 * The OpenKit-JavaScript version.
 * For now, it must be equal to the mobile agent version.
 */
exports.openKitVersion = '7.0.0000';
/**
 * Constant for the OneAgent Mobile Communication Protocol Version 3.
 */
exports.protocolVersion = 3;
/**
 * Constant for all OpenKit platforms.
 */
exports.platformTypeOpenKit = 1;
/**
 * The OpenKit-JavaScript agent technology type constant.
 */
exports.agentTechnologyType = 'okjs';
/**
 * The default server id, if no other has been chosen by the backend.
 */
exports.defaultServerId = 1;
/**
 * Error technology type "custom".
 */
exports.errorTechnologyType = 'c';
