// @dynatrace/openkit-js/dist/node/core/logging/ConsoleLogger.js
"use strict";
/*
 * Copyright 2019 Dynatrace LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const api_1 = require("../../api");
// tslint:disable:no-console
/**
 * Logger to log messages to the `console` object.
 *
 * @see {@link ConsoleLoggerFactory}
 */
class ConsoleLogger {
    constructor(name, logLevel) {
        this.name = `[${name}]`;
        this.level = logLevel;
    }
    debug(...msg) {
        if (this.isEnabled(api_1.LogLevel.Debug)) {
            console.debug(this.name, ...msg);
        }
    }
    info(...msg) {
        if (this.isEnabled(api_1.LogLevel.Info)) {
            console.info(this.name, ...msg);
        }
    }
    warn(...msg) {
        if (this.isEnabled(api_1.LogLevel.Warn)) {
            console.warn(this.name, ...msg);
        }
    }
    error(...msg) {
        if (this.isEnabled(api_1.LogLevel.Error)) {
            console.error(this.name, ...msg);
        }
    }
    isEnabled(level) {
        return this.level <= level;
    }
}
exports.ConsoleLogger = ConsoleLogger;
