// @dynatrace/openkit-js/dist/node/core/impl/null/NullSession.js
"use strict";
/*
 * Copyright 2019 Dynatrace LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const NullAction_1 = require("./NullAction");
const NullWebRequestTracer_1 = require("./NullWebRequestTracer");
class NullSession {
    enterAction(actionName) {
        return NullAction_1.defaultNullAction;
    }
    end() {
        // stub
    }
    identifyUser(userTag) {
        // stub
    }
    reportError(name, code, message) {
        // stub
    }
    reportCrash(errorName, reason, stacktrace) {
        // stub
    }
    traceWebRequest(url) {
        return NullWebRequestTracer_1.defaultNullWebRequestTracer;
    }
}
exports.NullSession = NullSession;
exports.defaultNullSession = new NullSession();
