// @dynatrace/openkit-js/dist/node/core/communication/http/state/HttpCommunicationChannel.js
"use strict";
/*
 * Copyright 2019 Dynatrace LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.HttpCommunicationChannel = void 0;
const SendingState_1 = require("./SendingState");
class HttpCommunicationChannel {
    constructor(client, loggerFactory) {
        this.logger = loggerFactory.createLogger('HttpCommunicationChannel');
        this.context = {
            stateMachine: this,
            loggerFactory,
            client,
        };
        this.state = new SendingState_1.SendingState(this.context);
    }
    sendNewSessionRequest(url, request) {
        return this.state.sendNewSessionRequest(url, request);
    }
    sendPayloadData(url, request, query) {
        return this.state.sendPayloadData(url, request, query);
    }
    sendStatusRequest(url, request) {
        return __awaiter(this, void 0, void 0, function* () {
            return this.state.sendStatusRequest(url, request);
        });
    }
    setNextState(channel) {
        this.state = channel;
    }
}
exports.HttpCommunicationChannel = HttpCommunicationChannel;
