// @dynatrace/openkit-js/dist/node/core/utils/CallbackHolder.js
"use strict";
/*
 * Copyright 2019 Dynatrace LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.CallbackHolder = void 0;
const Utils_1 = require("./Utils");
class CallbackHolder {
    constructor() {
        this.callbacks = [];
    }
    add(cb) {
        this.callbacks.push(cb);
    }
    remove(cb) {
        Utils_1.removeElement(this.callbacks, cb);
    }
    resolve(value) {
        this.callbacks.forEach((cb) => cb(value));
        this.callbacks.splice(0, this.callbacks.length);
    }
    contains(cb) {
        return this.callbacks.indexOf(cb) !== -1;
    }
}
exports.CallbackHolder = CallbackHolder;
