// @dynatrace/openkit-js/dist/node/core/utils/QueryBuilder.js
"use strict";
/*
 * Copyright 2019 Dynatrace LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.QueryBuilder = void 0;
const Utils_1 = require("./Utils");
class QueryBuilder {
    constructor() {
        this.options = {};
    }
    add(key, value, maxLength) {
        this.options[key] = Utils_1.truncate(value.toString(), maxLength);
        return this;
    }
    addIfDefined(key, value) {
        if (value !== undefined) {
            this.add(key, value);
        }
        return this;
    }
    addIfDefinedAndNotNull(key, value) {
        if (value !== undefined && value !== null) {
            this.add(key, value);
        }
        return this;
    }
    addIfNotNegative(key, value) {
        if (value >= 0) {
            this.add(key, value);
        }
        return this;
    }
    build() {
        return Object.keys(this.options)
            .map((key) => `${encodeURIComponent(key)}=${encodeURIComponent(this.options[key])}`)
            .join('&');
    }
}
exports.QueryBuilder = QueryBuilder;
